test/cpp/util/channelz_sampler.cc

```diff
@@ -0,0 +1,530 @@
+      std::queue<grpc::channelz::v1::Subchannel> subchannel_queue;
+      std::cout << "Tree depth = " << tree_depth << std::endl;
+      GetChannelDescedence(_topchannel, channel_queue, subchannel_queue);
+      while (!channel_queue.empty() || !subchannel_queue.empty()) {
+        ++tree_depth;
+        std::cout << "Tree depth = " << tree_depth << std::endl;
+        int ch_q_size = channel_queue.size();
+        int subch_q_size = subchannel_queue.size();
+        for (int i = 0; i < ch_q_size; ++i) {
+          grpc::channelz::v1::Channel ch = channel_queue.front();
+          channel_queue.pop();
+          GetChannelDescedence(ch, channel_queue, subchannel_queue);
+        }
+        for (int i = 0; i < subch_q_size; ++i) {
+          grpc::channelz::v1::Subchannel subch = subchannel_queue.front();
+          subchannel_queue.pop();
+          GetSubchannelDescedence(subch, channel_queue, subchannel_queue);
+        }
+      }
+    }
+  }
+
+  // dump data of all entities to stdout
+  void DumpStdout() {
+    std::string data_str;
+    for (const auto& _channel : all_channels_) {
+      std::cout << "channel " << GetChannelID(_channel)
+                << " data:" << std::endl;
+      ::google::protobuf::TextFormat::PrintToString(_channel.data(), &data_str);
+      printf("%s", data_str.c_str());
+    }
+    for (const auto& _subchannel : all_subchannels_) {
+      std::cout << "subchannel " << GetSubchannelID(_subchannel)
+                << " data:" << std::endl;
+      ::google::protobuf::TextFormat::PrintToString(_subchannel.data(),
+                                                    &data_str);
+      printf("%s", data_str.c_str());
+    }
+    for (const auto& _server : all_servers_) {
+      std::cout << "server " << GetServerID(_server) << " data:" << std::endl;
+      ::google::protobuf::TextFormat::PrintToString(_server.data(), &data_str);
+      printf("%s", data_str.c_str());
+    }
+    for (const auto& _socket : all_sockets_) {
+      std::cout << "socket " << GetSocketID(_socket) << " data:" << std::endl;
+      ::google::protobuf::TextFormat::PrintToString(_socket.data(), &data_str);
+      printf("%s", data_str.c_str());
+    }
+  }
+
+  // Store a channel in Json
+  void StoreChannelInJson(const grpc::channelz::v1::Channel& channel) {
+    std::string id = grpc::to_string(GetChannelID(channel));
+    std::string type = "Channel";
+    std::string description;
+    ::google::protobuf::TextFormat::PrintToString(channel.data(), &description);
+    StoreEntityInJson(id, type, description);
+  }
+
+  // Store a subchannel in Json
+  void StoreSubchannelInJson(const grpc::channelz::v1::Subchannel& subchannel) {
+    std::string id = grpc::to_string(GetSubchannelID(subchannel));
+    std::string type = "Subchannel";
+    std::string description;
+    ::google::protobuf::TextFormat::PrintToString(subchannel.data(),
+                                                  &description);
+    StoreEntityInJson(id, type, description);
```

Comment:
A little worried that the text-printed protobuf might not always parse as JSON
 
Can we instead generate the `description` variable used in `StoreEntityInJson` (here and elsewhere) via on the the protobuf JSON  APIs: https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/json-formatter

Reply:
Sorry this link refers to C# API? So if you look at the function StoreEntityInJson, we are in fact use grpc_core::Json::Object to do the parse. Does this release your worries?

Reply:
Even though `StoreEntityInJson` is properly creating a JSON object, it's `description` filed is still just a raw string as far as I can see. Can we perhaps use on of these APIs to do the  proto message to JSON string conversion: https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.util.json_util ?

Reply:
Sorry still a bit confused here. `description` should already be a JSON string instead of a raw string since it is updated by `::google::protobuf::TextFormat::PrintToString(const Message & message, std::string * output)` before `StoreEntityInJson`. (https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.text_format). We could use `google::protobuf::util::MessageToJsonString(const Message & message, std::string * output, const JsonOptions & options)`, but looks like the two functions are taking similar parameters?

Reply:
Convert raw string to Json done.